[FEATS][ORGANISMS][MÉTADONNÉES] Création d'organisme depuis les formulaires de CA et de JDD + Routes API [GET, POST]

VincentCauchois opened the pull request:
- [x] ⚠️ Faire le merge de https://github.com/PnX-SI/UsersHub-authentification-module/pull/130 et mise à jour du commit de UHAM ciblé ici, avant de merge la présente PR
- [-] ⚠️ Squash les commits de fixup avant merge
  - ❌ Pas fait finalement 

# DESCRIPTION

Implémente : 
- Une pop-in de création d'organisme intégrée au composant `pnx-metadata-actor`

Fonctionnalités :
- Création d'organisme depuis les forms de métadonnées :
  - Pop-in avec validation de formulaire et persistance des champs
  - Détection de noms d'organismes similaires par recherche approximative (distance de Levenshtein)
  - Liste paginée d'organismes similaires (3 par page) avec détails dépliables
  - Basé sur une nouvelle permission ["GEONATURE", "ORGANISM", "C"]
  - Rafraîchissement et sélection automatiques de l'organisme nouvellement créé
  - Traductions complètes en français et en anglais
- Ajout de routes API de gestion des organismes :
  - Récupération d'un organisme
  - Création d'un organisme

# EXEMPLES

### Création depuis le form de JDD

<img width="1920" height="1032" alt="0" src="https://github.com/user-attachments/assets/86a9d69e-29aa-4cbd-b88e-fa83c4182551" />

<img width="1920" height="1032" alt="1" src="https://github.com/user-attachments/assets/4ae2767b-43ca-4154-be09-b9825b6f8e5e" />

<img width="1920" height="1032" alt="2" src="https://github.com/user-attachments/assets/b30aa426-bb76-4afa-9e9a-0fce04fcde0f" />

<img width="1920" height="1032" alt="3" src="https://github.com/user-attachments/assets/5eaf1e18-3c19-4c08-906c-de78742ac4e9" />

<img width="1920" height="1032" alt="4" src="https://github.com/user-attachments/assets/8075d7bb-3f46-4337-b1e0-73a884527445" />

VincentCauchois committed:
feat(metadata): add organism creation buttons in af and ds forms

Implement organism creation dialog integrated into `pnx-metadata-actor` component.

Features:
- Material dialog with form validation and field persistence
- Duplicate organism name detection using fuzzy matching (Levenshtein distance)
- Paginated similar organisms list (3 per page) with expandable details
- Permission-based with ["METADATA", "ORGANISM", "C"]
- Auto-refresh and auto-select newly created organism
- Complete French and English translations

VincentCauchois committed:
fix(metadata): correct i18n paths for some texts

camillemonchicourt commented:
OK. Il aurait été préférable d'avoir un ticket décrivant une fonctionnalité comme celle-ci pour pouvoir en discuter et y faire référence au niveau conceptuel et fonctionnel.

- Dans ma compréhension, GeoNature ne devrait pas contenir de routes liés à des objets du schéma `utilisateurs` géré et centralisé dans UsersHub, donc s'appuyer sur les routes dans UsersHub.
- Pour les permissions, il faut y réfléchir en effet, mais pour moi les organismes sont des objets globaux et non pas des sous-objets de métadonnées
- Je parlerai de "Ajouter un organisme" et pas "Créer un organisme"
- A voir si on doit ajouter un paramètre pour permettre ou non d'accéder à cette fonctionnalité, ou si les permissions suffisent.

christophe-ramet commented:
> Dans ma compréhension, GeoNature ne devrait pas contenir de routes liés à des objets du schéma utilisateurs géré et centralisé dans UsersHub, donc s'appuyer sur les routes dans UsersHub.

La fonctionnalité de création d'organisme n'est pas vraiment liée à la gestion des utilisateurs. Typiquement, une instance qui utiliserait un autre provider que UsersHub peut aussi vouloir ajouter des organismes. On pourrait rajouter les fonctions permettant de créer des organismes dans Usershub-authentication-module, mais personnellement, je trouve ça un peu bizzare vu que ce n'est pas réellement lié aux Users ni à l'authentification. 

> Pour les permissions, il faut y réfléchir en effet, mais pour moi les organismes sont des objets globaux et non pas des sous-objets de métadonnées

Après réflexion, je te rejoins là-dessus. 

> A voir si on doit ajouter un paramètre pour permettre ou non d'accéder à cette fonctionnalité, ou si les permissions suffisent.

Ajouter un paramètre ça ferait un peu doublon non ? 


Pour la discussion autour des référentiels organismes, pour notre cas, ce n'est pas possible de se baser sur un référentiel. Dans une de nos instances, il ne peut exister de référentiels exhaustif. 
Du coup, la fonctionnalité de créer (ou d'activer) un organisme pré-existant dans un référentiel, pourrait avoir du sens, mais seulement **en plus** de celle de pouvoir les créer _from scratch_. Si c'est un besoin pour la communauté, ça pourrait faire l'objet d'un autre ticket/développement.

VincentCauchois commented:
> > OK. Il aurait été préférable d'avoir un ticket décrivant une fonctionnalité comme celle-ci pour pouvoir en discuter et y faire référence au niveau conceptuel et fonctionnel.
> > ```
> > * Dans ma compréhension, GeoNature ne devrait pas contenir de routes liés à des objets du schéma `utilisateurs` géré et centralisé dans UsersHub, donc s'appuyer sur les routes dans UsersHub.
> > 
> > * Pour les permissions, il faut y réfléchir en effet, mais pour moi les organismes sont des objets globaux et non pas des sous-objets de métadonnées
> > 
> > * Je parlerai de "Ajouter un organisme" et pas "Créer un organisme"
> > 
> > * A voir si on doit ajouter un paramètre pour permettre ou non d'accéder à cette fonctionnalité, ou si les permissions suffisent.
> > ```
> 
> Une autre réflexion qu'on aurait dû avoir, c'est la possibilité d'ajouter un organisme "depuis le référentiel organisme", et intégre ça à UsersHub directement : l'utilisateur à sa propre bib organisme comme maintenant, pas polluée, mais le référentiel organisme est stocké dans une table à part ert "propose" des organismes déjà connus dans le référentiels, en récupérant automatiquement les infos déjà disponibles dont l'uuid. Ca facilite la saisie, et ça évite qu'un organisme ait autant d'uuid dans le SINP que de partenaires qui diffusent leurs données...

On pourrait créer une nouvelle issue pour ça, ou voir même s'il n'existe pas déjà une telle issue.

En tout cas, et si l'on parle bien d'un référentiel des organismes tiers tel que celui du SINP, alors cela se pose de pouvoir récupérer tel référentiel. Ce n'est actuellement pas possible.

christophe-ramet reviewed (changes requested):
Ça me semble bien, hormis le commentaire dans la revue.

Il faudra aussi prendre en compte les remarques faites dans la PR.

VincentCauchois commented on frontend/src/app/metadataModule/organisms/organism-form-dialog.component.ts:
```diff
@@ -0,0 +1,264 @@
+  /**
+   * Calculate Levenshtein distance between two strings
+   */
+  private levenshteinDistance(str1: string, str2: string): number {
```
Cela impliquerait quand même:
- Un alourdissement de l'app par l'ajout de cette lib en entier
- Un risque et coût de maintenance en dépendant de cette lib

@christophe-ramet, @jacquesfize : vous pensez quoi de ces deux points ?

jacquesfize replied:
woah un distance de levenstein, ça fait longtemps ! Hum, j'opterai pour la lib. Comme dit @christophe-ramet, elle est très très légère. En plus, elle manipule uniquement des objets JS natifs, donc pas de dépendances supplémentaires.

VincentCauchois replied:
Ok, je vais partir sur la lib alors 👌 .

VincentCauchois committed:
fixup! feat(metadata): add organism creation buttons in af and ds forms

VincentCauchois committed:
fixup! feat(metadata): add organism creation buttons in af and ds forms

VincentCauchois committed:
fixup! feat(metadata): add organism creation buttons in af and ds forms

VincentCauchois committed:
fixup! feat(metadata): add organism creation buttons in af and ds forms

VincentCauchois commented on backend/geonature/core/users/routes.py:
```diff
@@ -227,8 +228,75 @@ def get_organismes_jdd():
+        "email_organisme": data.get("email_organisme"),
+        "url_organisme": data.get("url_organisme"),
+        "url_logo": data.get("url_logo"),
+    }
```
Ok, je vais voir pour ça.

VincentCauchois replied:
Vu en réunion, et on garde finalement le code concerné.

On s'assure ainsi que seuls les champs spécifiés sont retenus et notamment que ne peuvent être transmis les champs :
- `id_organisme`
- autres qui seraient éventuellement ajoutés dans le futur

VincentCauchois commented:
Ajoutée utilisation de la librairie `fastest-levenshtein` - https://www.npmjs.com/package/fastest-levenshtein

Et mise à jour des captures d'écran pour exemples dans la description de la PR.

VincentCauchois commented:
> > Dans ma compréhension, GeoNature ne devrait pas contenir de routes liés à des objets du schéma utilisateurs géré et centralisé dans UsersHub, donc s'appuyer sur les routes dans UsersHub.
> 
> La fonctionnalité de création d'organisme n'est pas vraiment liée à la gestion des utilisateurs. Typiquement, une instance qui utiliserait un autre provider que UsersHub peut aussi vouloir ajouter des organismes. On pourrait rajouter les fonctions permettant de créer des organismes dans Usershub-authentication-module, mais personnellement, je trouve ça un peu bizzare vu que ce n'est pas réellement lié aux Users ni à l'authentification.
> 
> > Pour les permissions, il faut y réfléchir en effet, mais pour moi les organismes sont des objets globaux et non pas des sous-objets de métadonnées
> 
> Après réflexion, je te rejoins là-dessus.
> 
> > A voir si on doit ajouter un paramètre pour permettre ou non d'accéder à cette fonctionnalité, ou si les permissions suffisent.
> 
> Ajouter un paramètre ça ferait un peu doublon non ?
> 
> Pour la discussion autour des référentiels organismes, pour notre cas, ce n'est pas possible de se baser sur un référentiel. Dans une de nos instances, il ne peut exister de référentiels exhaustif. Du coup, la fonctionnalité de créer (ou d'activer) un organisme pré-existant dans un référentiel, pourrait avoir du sens, mais seulement **en plus** de celle de pouvoir les créer _from scratch_. Si c'est un besoin pour la communauté, ça pourrait faire l'objet d'un autre ticket/développement.

Issue "[FEAT][ORGANISMES] Pouvoir se baser sur un ou plusieurs référentiels tiers d'organismes" ajoutée : https://github.com/PnX-SI/GeoNature/issues/3818

VincentCauchois committed:
fix(metadata-spec.js): fix some issues with metadata frontend tests

Fix:
- The test 'should delete the new "jeu de données"':
    - Fix the deletion of the JDD by interacting with confirmation dialog
    - Ensure the JDD is deleted by checking the absence of the JDD in the list
        - Search for the parent CA so as to include a remaining JDD to be present and help verifying the loading of the list of JDD before verifying the absence of the removed JDD
    - Avoid the use of `cy.wait` with arbitrary timeouts by using `.should('be.visible')` and `cy.intercept`
    - Add some comments
- The test 'should delete the new "cadre d\'acquisition"':
  - Avoid the use of `cy.wait` with arbitrary timeouts by using `.should('be.visible')` and checks for the dialog to be closed and the CA to no longer be present
  - Ensure the CA is deleted by checking its absence in the list
- Some typographic errors

VincentCauchois committed:
feat(metadata-spec.js): add frontend tests create organism from metadata

Notes:
- Also test the add of a second main contact for a JDD, which was previously untested
- Important: the new organism is deleted, to ensure the data would not persist through tests and operations in a local database, and thus it requires the feature "delete organisms from GN admin" implemented in a previous commit
- Important: the delete of the organism is made after the delete of the JDD which it is associated to, to prevent from the association blocking the delete

VincentCauchois committed:
fix(submodule): set "usershub-auth-module" to current develop

VincentCauchois committed:
fixup! feat(metadata): add organism creation buttons in af and ds forms

VincentCauchois committed:
feat(pytest): add backend tests get, create and delete organism

PyTest tests for new routes:
- `get_organism`
- `create_organism`
- `delete_organism`

VincentCauchois commented:
> * [x]  Ajout de tests BE

Tests BE ajoutés // commit "feat(pytest): add backend tests get, create and delete organism"

jacquesfize committed:
fix rebase

jacquesfize committed:
fix backend test

jacquesfize committed:
refact create organism button

jacquesfize committed:
remove delete route since it is not used

jacquesfize committed:
bump usershub-a-m

jacquesfize committed:
fix frontend test

jacquesfize committed:
fix frontend test 2

jacquesfize committed:
fix lint

VincentCauchois commented on frontend/cypress/e2e/metadata-spec.js:
```diff
@@ -73,7 +73,7 @@ describe('Testing metadata', () => {
     cy.get('[data-qa="pnx-organism-form-dialog-name"]').type(newOrganism.name);
     //    Verify that the organism 'ma structure test' is displayed in the list of similar organisms
-    cy.get('[data-qa="pnx-organism-form-dialog-similar-list"] li').should('have.length', 1);
+    cy.get('[data-qa="pnx-organism-form-dialog-similar-list"] li').should('have.length.least', 1);
```
@jacquesfize : normalement on devrait avoir exactement 1 organisme dans la liste, et sinon c'est qu'il y a un problème quelque part. Ne serait-ce pas de garder un test d'égalité stricte ?

jacquesfize committed:
move similarity function to /organisms

jacquesfize committed:
remove levenstein package
